File: ValidParentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = list()
        #mapping table
        table = {
            ')':'(',
            '}':'{',
            ']':'[',
        }
        
        for c in s:
            if c not in table: #find c as a key 
                stack.append(c) #c is not key then append to the stack
            elif not stack or table[c] != stack.pop(): #if the stack is empty 
                                                       #or the value mapping to c is not equal to the last inserted element of the stack
                return False                           #return false
        
        return len(stack)==0 #if the stack is empty after the loop -> true

File: ValidParentheses.java
class Solution {
    public boolean isValid(String s) {
        //char can be ( ) { } [] ;
        //stack
        //pop only if char == )
        //check the stack if it's empty 
        //empty -> t, otherwise -> f
        
        Stack<Character> stack = new Stack<>();

        for(int i=0;i<s.length();i++){
            
            char token = s.charAt(i);
            
            if(token == ')' || token == '}' || token == ']'){
                
                if(stack.isEmpty()) return false; 
                
                if(token == ')' && stack.peek() == '('){
                    stack.pop();
                }
                else if(token == '}' && stack.peek() == '{'){
                    stack.pop();
                }
                else if(token == ']' && stack.peek() == '['){
                    stack.pop();
                }
                else return false;
            }
            
            else stack.push(token);
        }
        
        return stack.isEmpty();
        
//         HashMap<Character, Character> closeToOpen = new HashMap<>();
        
//         closeToOpen.put(')','(');
//         closeToOpen.put('}','{');
//         closeToOpen.put(']','[');
        
//         for(int i=0;i<s.length();i++){
//             if(closeToOpen.containsKey(s.charAt(i))){ //close symbol 
//                 //validate if the close symbol is in correct order and pair
//                 if(!stack.isEmpty() && stack.peek() == closeToOpen.get(s.charAt(i))) {
//                     stack.pop();
//                 }
//                 else{
//                     return false;  
//                 }
//             }
//             else{
//                     stack.push(s.charAt(i));    
//             }
            
//         }
        
//         return stack.isEmpty();
    }
}
